metadata-builders/src/main/java/com/linkedin/metadata/builders/search/MLModelIndexBuilder.java

```diff
@@ -0,0 +1,203 @@
+            .replace('.', '/').toLowerCase();
+    }
+
+    /**
+     * Given model urn, this returns a {@link MLModelDocument} model that has urn, dataset name, platform and origin fields set
+     *
+     * @param urn {@link MLModelUrn} that needs to be set
+     * @return {@link MLModelDocument} model with relevant fields set that are extracted from the urn
+     */
+    @Nonnull
+    private static MLModelDocument setUrnDerivedFields(@Nonnull MLModelUrn urn) {
+        return new MLModelDocument()
+            .setName(urn.getMlModelNameEntity())
+            .setOrigin(urn.getOriginEntity())
+            .setPlatform(urn.getPlatformEntity().getPlatformNameEntity())
+            .setUrn(urn)
+            .setBrowsePaths(new StringArray(Collections.singletonList(buildBrowsePath(urn))));
+    }
+
+    @Nonnull
+    private List<MLModelDocument> getDocumentsToUpdateFromSnapshotType(@Nonnull MLModelSnapshot mlModelSnapshot) {
+        final MLModelUrn urn = mlModelSnapshot.getUrn();
+        return mlModelSnapshot.getAspects().stream().map(aspect -> {
+            if (aspect.isCaveatsAndRecommendations()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getCaveatsAndRecommendations());
+            } else if (aspect.isEthicalConsiderations()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getEthicalConsiderations());
+            } else if (aspect.isEvaluationData()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getEvaluationData());
+            } else if (aspect.isInstitutionalMemory()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getInstitutionalMemory());
+            } else if (aspect.isIntendedUse()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getIntendedUse());
+            } else if (aspect.isMetrics()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getMetrics());
+            } else if (aspect.isMLModelFactorPrompts()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getMLModelFactorPrompts());
+            } else if (aspect.isMLModelProperties()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getMLModelProperties());
+            } else if (aspect.isOwnership()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getOwnership());
+            } else if (aspect.isQuantitativeAnalyses()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getQuantitativeAnalyses());
+            } else if (aspect.isStatus()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getStatus());
+            } else if (aspect.isTrainingData()) {
+                return getDocumentToUpdateFromAspect(urn, aspect.getTrainingData());
+            }
+            return null;
+        }).filter(Objects::nonNull).collect(Collectors.toList());
```

Comment:
There are bunch of `getDocumentToUpdateFromAspect` methods where you are essentially setting urn parts - nothing to do with the aspect. I suggest we get rid of those methods and instead just add `documents.add(setUrnDerivedFields(urn));` after line 92.

Let me know your thoughts.

Reply:
That should be fine, if a snapshot comes in with an aspect that isn't a part of MLModelAspect then it should fail in serialization or ignore it so shouldn't cause any issues.

Reply:
I am not worried about an aspect that isn't part of `MLModelAspect` but if in future someone adds a new aspect to `MLModelAspect` we should at least index the urn parts when this new aspect is received w/o having to modify the index builder logic to process this new aspect.

I have tried to explain the same in this PR https://github.com/linkedin/datahub/pull/1937

Reply:
I'll update to follow that pattern. One thing I'm curious about, why is this returning a list of documents rather than a single one with all the fields set? It seems like they have the same end effect, but is having multiple document objects that only update a few fields preferable to a single one that updates them all?

Reply:
Agree, single document is better than multiple documents (with the same end effect). However these documents can come from processing multiple aspects and will therefore require merging of documents into one. Also, once we move to aspect specific events, we should have one search document per event.